### PR TITLE
fix(desktop): 防止 Hook 请求在更新重启后重复执行

### PR DESCRIPTION
**Pull request by dashhuang**
### 摘要

更新重启后，服务端可能在重连时重投已经完成的 `requestId`。Desktop 原先只用进程内 `ackHistory` 去重，重启会清空它，导致 Slack 和 X 共用的 Hook dispatcher 再次调用 Agent。

本 PR 为客户端增加 owner-scoped 的持久终态账本与 pending outbox：完成的请求在重启后回放原 ACK 与文本终态，不再重复执行；明确未送达的终态会在重连时补发。跨重启账本只保存文本终态，不保存 base64 附件字节。

### 变更类型

- [x] `fix` 缺陷修复
- [ ] `feat` 新功能
- [ ] `refactor` / `perf` 重构或性能优化
- [ ] `docs` / `test` / `chore` 文档、测试或工程维护
- [ ] 其他：

### 范围

- 关联 Issue / 需求：无
- 本 PR 包含：Desktop Hook dispatcher 的跨重启去重、终态 outbox、Slack/X 共用连接的持久化装配，以及回归测试。
- 明确不包含：服务端改动、wire protocol 改动、附件持久化、其它客户端或 UI 改动。
- 用户可见变化：更新重启后，Slack/X 不会因为服务端重投已完成请求而再次执行并重复回答；重启前已完成但尚未送出的文本终态会在重连后补发。
- 是否存在 breaking change：无

### UI 变化

- 不涉及：仅修改 Desktop main 侧 Hook dispatcher、owner-scoped 本地账本和测试，没有 UI、交互或文案变化。
- 引用的设计规范：不涉及。

### 自动验证

```text
pnpm --filter desktop exec vitest run src/main/hook-control/__tests__/requestLedger.test.ts src/main/hook-control/__tests__/dispatcher.test.ts
结果：117 tests passed

pnpm --filter desktop exec vitest run src/main/hook-control/__tests__
结果：466 tests passed

pnpm --filter desktop run --if-present typecheck
结果：通过

bash /Users/dash/Code/XD/dash/Skills/git/scripts/run-unit-gate.sh /Users/dash/Code/Cindy/cindy-fix-hook-replay-after-restart
结果：GATE_EXIT=0（rebase origin/main 后最终 HEAD 已重新执行）
```

### 手工验证

不涉及：本 PR 未启动 Desktop，也未使用真实 Slack/X 账号做端到端发送；行为由 dispatcher/ledger 回归测试覆盖。

### 未执行的验证

未执行真实服务端重连和应用层 receipt 验证。当前协议的 `ws.send() === true` 只表示本地 WebSocket 已接受入队，不代表服务端已经处理 `turn.end`；在“本地 enqueue 成功、服务端处理前进程崩溃”的窗口，客户端无法仅靠现有 wire protocol 严格证明送达，需要后续服务端 receipt/协议支持。

### 风险分类

- [ ] 无已知风险
- [ ] SQLite / migration
- [ ] system prompt
- [x] 协议兼容
- [x] 权限 / 安全 / 用户数据
- [ ] 存量插件兼容（批准状态 / 指纹 / manifest 校验 / 安装布局 / 包格式）
- [ ] 原生层 / fingerprint / OTA
- [ ] 跨平台差异
- [x] 其他：本地可靠性与重连恢复

### 影响与回滚

- 影响范围：Desktop Hook 的 Slack/X 请求分发；新增文件位于 owner-scoped userData 下，账本限制为最多 2,000 条、总文件大小约 32 MB，写入失败时降级到原有进程内去重，不吞掉当前回答。
- 回滚 / 降级方式：回滚本 PR 即恢复原 dispatcher 行为；新建的 `hook-request-ledger.json` 不参与其它功能，旧版本会忽略它。未修改 wire protocol，不需要服务端同步升级。

### 提交前检查

- [x] 已 review 完整 diff
- [x] 每个 commit 都带 DCO 签名（`git commit -s`，见 [DCO](../DCO)）
- [x] UI 改动已在「UI 变化」注明引用的设计规范章节（不涉及 UI，已注明）
- [x] 未提交凭证、令牌或授权文件
- [x] 已补充必要文档
- [x] 已确认测试结果或说明未执行原因

### CODE REVIEW

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 3 条 review conversation 没 resolve(apps/desktop/src/main/hook-control/requestLedger.ts / apps/desktop/src/main/hook-control/dispatcher.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: apps/desktop/src/main/hook-control/requestLedger.ts [hunk 1]
```diff
@@ -0,0 +1,237 @@
...
+    }
+    if (raw === null) {
+      cachedEntries = [];
+      return cachedEntries;
+    }
+    if (raw.length > maxFileChars) {
+      deps.log.warn('read hook request ledger failed (file-too-large)');
+      cachedEntries = [];
+      return cachedEntries;
+    }
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/hook-control/requestLedger.ts]: 已修复于 1e7515c3b：账本现在用 Buffer.byteLength(value, utf8) 统一按 UTF-8 字节数限制原始文件、单条记录和序列化后的总文件大小；回归测试增加了多字节 emoji 文件场景。

File: apps/desktop/src/main/hook-control/dispatcher.ts [hunk 2]
```diff
@@ -1950,13 +2059,34 @@ export function createHookDispatcher(deps: HookDispatcherDeps): HookDispatcher {
...
+      } catch {
+        log.warn('hook durable turn.end outbox lookup threw; waiting for request replay');
+      }
+      for (const pending of durablePending) {
+        // The full in-memory frame was already accepted above. If updating the
+        // ledger failed, do not immediately duplicate it with the text-only
+        // durable frame during the same reconnect attempt.
+        if (flushedRequestIds.has(pending.requestId)) continue;
+        if (!pending.turnEnd || !send(makeTurnEnd(pending.turnEnd))) return;
+        markTerminalSent(connectionId, pending.requestId);
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/hook-control/dispatcher.ts]: 已修复于 1e7515c3b：重连补发 durable turn.end 成功后，如果 markSent 返回 false，现在会回退调用 persistTerminal(... delivery: sent)；新增回归测试验证账本状态不会一直停在 pending，避免后续重连重复补发。

File: apps/desktop/src/main/hook-control/dispatcher.ts [hunk 3]
```diff
@@ -1572,8 +1648,34 @@ export function createHookDispatcher(deps: HookDispatcherDeps): HookDispatcher {
...
+      log.warn('hook terminal request lookup threw; using in-memory dedupe only');
+    }
+    if (terminalReplay) {
+      cacheAck(connectionId, terminalReplay.ack);
+      const ackDelivered = send(makeTaskAck(terminalReplay.ack));
+      if (
+        ackDelivered &&
+        terminalReplay.delivery === 'pending' &&
+        terminalReplay.turnEnd &&
+        send(makeTurnEnd(terminalReplay.turnEnd))
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/hook-control/dispatcher.ts]: 已修复于 1e7515c3b：收到持久账本中的已完成 requestId 重投时，先回放原 ACK，再回放去附件后的 turn.end 文本终态，无论本地 delivery 是 pending 还是 sent；不会重新调用 runner。服务端现有 requestId 幂等语义负责吸收可能的重复终态帧，本 PR 未修改服务端或 wire protocol。新增测试覆盖完成请求和排队取消请求的跨 dispatcher 重投。

**Review by MagicLizi — APPROVED** [review]
代码审查通过。at-least-once + 幂等 requestId + durable outbox 组合正确;所有存储故障路径 fail-open 不吞消息;eviction 不踢 pending 项;跨实例/跨 connection 隔离;损坏文件重建。

**Comment by MagicLizi** [comment]
durable outbox + 幂等重放的设计让更新重启后不丢消息也不重复执行,fail-open 降级路径也经得起推敲。审查通过但本轮因结构性检查门未能合并,下轮重试。

**Comment by dashhuang** [comment]
维护说明:已通过 merge commit `3d86e17c1` 解决与 main(#1794 turn.delivery 客户端)在 `dispatcher.ts` 的冲突。组合语义(最小组合,不做融合重构):

- **账本(跨重启)与 ACK 内存队列(会话内)共存**:各自负责自己维度的重发;
- **唯一接线点**:X 连接协商 `turn-delivery-v1` 时,账本记录保持 `pending`,收到任一 `turn.delivery` 回执才收口为 `sent`(`handleTurnDelivery` 中 `markTerminalSent`)。这补上了 #1794 声明的「accepted 前进程崩溃」剩余边界:重启后由账本经 ACK 缓冲补发,服务端幂等由 server 侧收据(cindy-server#328)兜底;
- **重连补发去重**:`onConnected` 中账本补发跳过已在 ACK 缓冲中的 requestId,避免同一终态的文本帧重复发送;
- 旧服务端(未协商)完全保持本 PR 原有 ws.send 判据语义,Slack 渠道不受影响。

新增 2 个交错回归测试(ACK 模式账本收口、回执前重启补发)。验证:hook-control 475 tests 全过、desktop typecheck 过、根 `pnpm test:unit` 门禁 GATE_EXIT=0。

融合重构(单一送达状态机)与 #1794 遗留 P2(降级回落遗留 timer)保持在本 PR 之外,作为合并后的独立 follow-up(见 #1794 评论)。

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 1 条 review conversation 没 resolve(apps/desktop/src/main/hook-control/dispatcher.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: apps/desktop/src/main/hook-control/dispatcher.ts [hunk 4]
```diff
@@ -1661,8 +1744,41 @@ export function createHookDispatcher(deps: HookDispatcherDeps): HookDispatcher {
...
+      }
+      if (!send(makeTurnEnd(terminalReplay.turnEnd))) {
+        persistTerminalRecord({ ...terminalReplay, delivery: 'pending' });
+        return;
+      }
+      if (
+        terminalReplay.delivery === 'pending' &&
+        !markTerminalSent(connectionId, payload.requestId)
+      ) {
+        persistTerminalRecord({ ...terminalReplay, delivery: 'sent' });
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/hook-control/dispatcher.ts:1786]: 已修复于 91d079b5f:`supportsDeliveryAck(connectionId)` 为 true 时,重投回放的 turn.end 改走 `trackPendingDelivery()`(与 `sendOrBuffer()` 同语义),账本保持 `pending` 直到 `handleTurnDelivery` 收到回执才收口;旧记录为 `sent` 时降回 `pending`(server 重投即说明无持久收据)。新增回归测试「ACK 模式重投终态 -> 回放经 ACK 缓冲退避重发, 账本保持 pending 直到回执」钉住该行为。

**Review by MagicLizi — APPROVED** [review]
Heavy review passed. Well-designed durable terminal state ledger that prevents duplicate hook execution after auto-update restart. No security issues, no malicious patterns, comprehensive test coverage for all edge cases including disk failures and partial delivery. Only P2-level efficiency observations.

**Comment by MagicLizi** [comment]
Hook 请求的幂等难题终于有了持久化方案——用户升级重启后不再莫名重跑已完成的任务,ledger 的 eviction 策略和磁盘容错逻辑都很稳当。
